### DISCUSSION TITLE
Wind Direction Units in Point Stat on WCOSS2

### DISCUSSION THREAD

**Discussion by johnlwagner** [dtcenter/METplus#1868] [Statistical Computation] [answered]
Greetings
I believe I have having issues with the wind direction units in point_stat on WCOSS2.  I believe this issue is similar to one I had a few months ago: https://github.com/dtcenter/METplus/discussions/1744.
When I run point stat, I don't believe I am using the correct units for the u and v components.  When I run stat_analysis, the MAEs for a month are typically around 90 degrees for all stations and time projections.  For short term projections, I would expect that number to be closer to 20 degrees.
The issue most likely is how I am computing the u and v components from the wind speed (knots) and direction.  First, for the METAR values, the wind speed is converted from knots to meters per second (model grib2 inputs are already in meters per second).  I am not having any issues with my wind speed verification, so I don't believe there is an issue here.
The components are computed using numpy with these formulas, as they were on WCOSS1:
u = -ws*numpy.sin(wd)
v = -ws*numpy.cos(wd)
I am testing now using radians when computing the components, as I thought the units may have changed on WCOSS2 (like they did with gridded wind direction before I started specifying the GRIB2_center):
u = -ws*numpy.sin(wd*(numpy.pi/180))
v = -ws*numpy.cos(wd*(numpy.pi/180))
I am not sure though how Point Stat handles the units internally.  Any insights or advice on how to calculate the wind components on WCOSS2 before calling Point Stat would be appreciated.
Thanks
John

**Comment by johnlwagner** [comment, marked as answer]
I guess I should have checked the preview, as the *s in my equations added italics instead.
I'll use an "x" to mean multiply here:
Equations from WCOSS1:
u = -ws x numpy.sin(wd)
v = -ws x numpy.cos(wd)
Test equations on WCOSS2 using radians:
u = -ws x numpy.sin(wd x (numpy.pi/180))
v = -ws x numpy.cos(wd x (numpy.pi/180))

**Reply by JohnHalleyGotway** [reply]
Hi @johnlwagner I see you have a question about wind vectors in MET.

As you already know, MET expects the inputs to contain the U and V wind components. And it can derive wind speed and direction on the fly from the U and V components by calling the [convert_u_v_to_wdir()](https://github.com/dtcenter/MET/blob/develop/src/basic/vx_math/angles.cc#L181) and [convert_u_v_to_wind()](https://github.com/dtcenter/MET/blob/6f498e1fba51bb1fc9156bc9aafd5f7fa33e5ba9/src/basic/vx_math/angles.cc#L203) functions.

You have observations of wind speed and direction and would like to derive the U/V components from them via Python.

Looks like [numpy.sin](https://numpy.org/doc/stable/reference/generated/numpy.sin.html#numpy.sin) and [numpy.cos](https://numpy.org/doc/stable/reference/generated/numpy.cos.html#numpy.cos) do process inputs in radians rather than degrees. So you're correct to convert the direction from degrees to radians. And multiplying by `numpy.pi/180` should do it. You could chain in the [numpy.radians](https://numpy.org/doc/stable/reference/generated/numpy.radians.html) function but either should yield the same result.

Looking at your equations, I have 2 thoughts:
1. I think you have the sine and cosine reversed. Checking some [websites](http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv), I think it should be:
```
u = ws * numpy.cos(wd)
v = ws * numpy.sin(wd)
```
2. I'm not sure about multiplying by `-1`. It just depends on how the observation of wind direction is defined. Is it "direction from" or "direction to"... and does that match the output being generated by the model.

Fortunately, 2. is pretty easy to check. If you see errors around 180 degrees, you probably have it exactly wrong and can remove the -1.

Please try swapping the sin and cos and let me know how it goes.

Thanks!

**Reply by johnlwagner** [reply]
Thanks John.  My initial test with your changes looks good, so I'm running more tests.
I'm not sure who wrote our code for this, but like me, they may have been misled by the ECMWF page on this: https://confluence.ecmwf.int/pages/viewpage.action?pageId=133262398
As always, I appreciate your help.
John

**Reply by JohnHalleyGotway** [reply]
Wow, that link is very confusing. I also note that their code calls `atan2(v,u)` where [this line](https://github.com/dtcenter/MET/blob/011f6abf9fd2ccf80d56e521fa4052771412275d/src/basic/vx_math/angles.cc#L191) in MET calls `atan2d(-u,-v)`. So the order and sign of u and v are reversed.

I can't say confidently enough that that info is incorrect, but it is certainly doesn't follow other conventions I've seen.

Thanks for sending the link.

**Reply by JohnHalleyGotway** [reply]
@johnlwagner just following up on this discussion. Do any issues or question for METplus remain about your derivation of U and V from wind speed and direction?

**Comment by johnlwagner** [comment]
I believe everything is working correctly now.  Thanks John!
